### ISSUE TITLE
stop_continuous_recognition() inconsistent delays

### ISSUE THREAD

**Issue by EnzyCue** [Azure-Samples/cognitive-services-speech-sdk#1506]
I am using continuous recognition to recognise audio from a microphone but I want to be able to switch it on and off without delay. The problem I am facing is that whenever I call stop_continuous_recognition() I get a 2-10 second complete halt of my entire program.  

I have tried experimenting with threading as well as switching to async and while I have been able to achieve some success it is usually very inconsistent, i.e there are times where I would get it to have no delay whatsoever but I would get a 10 second blocking of the main thread every 5 or so consecutive usages. I have experienced this inconsistency with and without using the async versions so I don't think that's the problem as I have seen in some other posts on here. 

I am pretty confused and I would love to get some explanation as to how these stop recognition functions exactly work. I have also thought about experimenting with creating my own Connection object but I had no success there as well.

I currently suspect that the halting is triggered when the stop function is called while a connection process is happening behind the scenes. But obviously I cant be sure and neither do I know a good of way of getting around that right now. 

I would appreciate any help I could get on this!

code:

```Python
event_start_recognition = threading.Event()
event_stop_recognition = threading.Event()


def start_recognition():
    event_start_recognition.set()


def stop_recognition():
    event_stop_recognition.set()


class MySpeechRecognizer:
    def __init__(self):        
        speech_config = speechsdk.SpeechConfig(subscription=subsKey, region="australiaeast")
        speech_config.speech_recognition_language = "en-US"

        audio_config = speechsdk.audio.AudioConfig(use_default_microphone=True)

        self.speech_recognizer = speechsdk.SpeechRecognizer(speech_config=speech_config, audio_config=audio_config)
        self.initialise_connections()
        

    def initialise_connections(self):
        self.speech_recognizer.recognized.connect(lambda evt: self.handle_result(evt))
        self.speech_recognizer.recognizing.connect(lambda evt: print('RECOGNIZING: {}'.format(evt)))


    def handle_result(self, evt):
        print(evt.result.text)

    def start_up(self):
        event_start_recognition.wait()
        self.speech_recognizer.start_continuous_recognition_async()
        event_stop_recognition.wait()
        self.speech_recognizer.stop_continuous_recognition_async()


if __name__ == "__main__":
    speech_recogniser = MySpeechRecognizer()

    thread = threading.Thread(target=speech_recogniser.start_up)
    thread.start()

    input("Press a key to start recognition...\n")
    start_recognition()
    input("Press a key to stop recognition...\n")
    stop_recognition()

    thread.join()
```

**Comment by ralph-msft** [comment]
Thanks for using the Cognitive Services Speech SDK!

I was able to reproduce the issue you were seeing with the 2-10 second stalls. It looks like there's is an odd quirk where if you don't wait on or use the future returned from the stop_continuous_recognition_async(), it causes some stalls in the SDK.

As a workaround you can switch to using the synchronous versions of the start and stop functions from your background thread. I've modified your code above to test this and added a loop so you can start/stop multiple times. I can no longer reproduce the 2-10 second stalls with these changes:

```Python
event_start_recognition = threading.Event()
event_stop_recognition = threading.Event()

def start_recognition():
    event_start_recognition.set()


def stop_recognition():
    event_stop_recognition.set()


class MySpeechRecognizer:
    def __init__(self):        
        speech_config = speechsdk.SpeechConfig(subscription=subsKey, region=subsRegion)
        speech_config.speech_recognition_language = "en-US"

        audio_config = speechsdk.audio.AudioConfig(use_default_microphone=True)

        self.speech_recognizer = speechsdk.SpeechRecognizer(speech_config=speech_config, audio_config=audio_config)
        self.initialise_connections()
        

    def initialise_connections(self):
        self.speech_recognizer.recognized.connect(lambda evt: self.handle_result(evt))
        self.speech_recognizer.recognizing.connect(lambda evt: print('RECOGNIZING: {}'.format(evt)))
        self.speech_recognizer.canceled.connect(lambda evt: print('CANCELED: {}'.format(evt)))

    def handle_result(self, evt):
        print(evt.result.text)

    def start_up(self):
        while True:
            event_start_recognition.wait()
            self.speech_recognizer.start_continuous_recognition()
            event_start_recognition.clear()

            event_stop_recognition.wait()
            self.speech_recognizer.stop_continuous_recognition()
            event_stop_recognition.clear()


if __name__ == "__main__":
    speech_recogniser = MySpeechRecognizer()

    thread = threading.Thread(target=speech_recogniser.start_up)
    thread.start()

    while True:
        input("Press a key to start recognition...\n")
        start_recognition()

        input("Press a key to stop recognition...\n")
        stop_recognition()
```

**Comment by EnzyCue** [comment]
@ralph-msft thanks for the reply! 

I tried out your code and it is definitely an improvement but sometimes if I cancel the call when it hasn't completed recognition it will cause the delay again. Is there to get around this or is that not possible to work around?

**Comment by EnzyCue** [comment]
Sure. I have attached the log file for me saying "ok, I will experience the delay now" but ending the recognition before service can recognise the entire utterance (thereby causing the delay).

[Logs.txt](https://github.com/Azure-Samples/cognitive-services-speech-sdk/files/8883789/Logs.txt)

**Comment by EnzyCue** [comment]
I am happy to carry out more tests and send through the logs if you need them. Let me know if you need anything like that.

**Comment by adammpolak** [comment]
Was this ever solved?

**Comment by pankopon** [comment]
Closing the item since no update from @ralph-msft in 6 months. Please re-test with the latest Speech SDK and open a new issue if still necessary.
